### PR TITLE
[structural_application] update to compile with pybind and amatrix

### PR DESCRIPTION
**Pull request by vryy**
open a new pull request and closed the old one

### CODE REVIEW

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 1]
```diff
@@ -1097,7 +1097,7 @@ class ResidualBasedBlockBuilderAndSolver
 
             TSystemVectorType bmodified(rb.size());
             TSparseSpace::Mult(T_transpose_matrix, rb, bmodified);
-            noalias(rb) = bmodified;
+            TSparseSpace::Copy(bmodified, rb);
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: changes in the core, esp in such a class MUST go to a separate PR

File: kratos/includes/amatrix_interface.h [hunk 2]
```diff
@@ -609,6 +609,19 @@ typename TExpression1Type::data_type inner_prod(
 
 }
 
+    template <typename TExpressionType, std::size_t TCategory>
+    typename TExpressionType::data_type norm_1(
```
Comment (philbucher) [inline kratos/includes/amatrix_interface.h:613]: same
also use `std::abs` instead of `std::fabs`

**Review by philbucher — CHANGES_REQUESTED** [review]
Several things:
- changes to core-stuff must be in a separate PR, those changes affect everyone
- please do not commit files from GiD, e.g. msh, prb, vv etc, the repository is already large enough. If you want to add examples then add them to the Examples-repository
- this PR is gigantic, experience shows that reviewing takes very long, so I suggest to make smaller PRs in the future. For me this is not a blocker though since it only affects your app

File: kratos/includes/amatrix_interface.h (hunk 2 above)
Comment (vryy) [reply kratos/includes/amatrix_interface.h:613]: why? std::abs works with only int and long

File: applications/structural_application.orig/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,139 @@
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
```
Comment (RiccardoRossi) [inline applications/structural_application.orig/CMakeLists.txt]: i have the feeling you are committing from a hidden directory (with the .orig in the name)

File: applications/structural_application/constitutive_laws/brittle_material_2d.cpp [hunk 4]
```diff
@@ -216,8 +216,9 @@ void BrittleMaterial2D::FinalizeSolutionStep( const Properties& props,
         const Vector& ShapeFunctionsValues ,
         const ProcessInfo& CurrentProcessInfo)
 {
+    double detF;
     noalias(mStrain_Old) = mStrain_New;
-    SD_MathUtils<double>::InvertMatrix(mF,m_invF_old);   /// WARNING = descomeatr
+    MathUtils<double>::InvertMatrix(mF,m_invF_old,detF);   /// WARNING = descomeatr
```
Comment (RiccardoRossi) [inline applications/structural_application/constitutive_laws/brittle_material_2d.cpp]: here you could remove the comment

File: applications/structural_application/custom_strategies/strategies/residualbased_arc_length_strategy.h [hunk 5]
```diff
@@ -53,35 +53,33 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
...
-
+#include <cmath>
 
 /* External includes */
 
 /* Project includes */
 #include "structural_application.h"
 #include "includes/define.h"
 #include "includes/model_part.h"
+#include "containers/model.h"
```
Comment (RiccardoRossi) [inline applications/structural_application/custom_strategies/strategies/residualbased_arc_length_strategy.h:64]: I think this file is included in the core. could u take a look?

**Review by RiccardoRossi — COMMENTED** [review]
@vryy,

the most important comment is that i think you wrongly committed a backup directory, please have a look before we take a more in depth review

File: applications/structural_application/custom_strategies/strategies/residualbased_arc_length_strategy.h (hunk 5 above)
Comment (RiccardoRossi) [reply applications/structural_application/custom_strategies/strategies/residualbased_arc_length_strategy.h:64]: (i refer to the arc_lenght_strategy)

Comment (vryy) [reply applications/structural_application/custom_strategies/strategies/residualbased_arc_length_strategy.h:64]: Hi @RiccardoRossi , the residualbased_arc_length_strategy.h (or similar) is not yet in the core

File: applications/structural_application.orig/CMakeLists.txt (hunk 3 above)
Comment (vryy) [reply applications/structural_application.orig/CMakeLists.txt]: thanks, I removed it in the latest commit

**Comment by philbucher** [comment]
there are still changes in the core, I think you should merge master
After this is done I will review

**Comment by vryy** [comment]
I have a problem with merge because it falls too far behind the head and impossible to fix individual incremental merge. I will close this branch and make new request.

**Comment by philbucher** [comment]
is this branch still needed?

**Comment by vryy** [comment]
no, just delete it
